File: packages/python/src/a2a_lite/llm.py
"""
LLM provider integration decorators for A2A Lite.

Provides decorators that wrap skills to call LLM APIs (OpenAI, Anthropic, Ollama).
Each uses optional import patterns — the LLM library is only required at runtime.

Requires:
    pip install a2a-lite[openai]     # for openai_skill
    pip install a2a-lite[anthropic]  # for anthropic_skill
    ollama_skill uses httpx (already a core dep)

Example:
    from a2a_lite.llm import openai_skill, anthropic_skill

    @agent.skill("chat", streaming=True)
    @openai_skill(model="gpt-4o-mini", system_prompt="You are helpful.")
    async def chat(message: str) -> str:
        ...  # handled by decorator

    @agent.skill("analyze")
    @anthropic_skill(model="claude-sonnet-4-5-20250929")
    async def analyze(text: str) -> str:
        ...  # handled by decorator
"""

from __future__ import annotations

import functools
from typing import Any, Callable, Optional


def openai_skill(
    model: str = "gpt-4o-mini",
    system_prompt: str = "You are a helpful assistant.",
    temperature: float = 0.7,
    max_tokens: Optional[int] = None,
    streaming: bool = False,
    **extra_kwargs: Any,
) -> Callable:
    """Decorator that wraps a skill to call the OpenAI chat completions API.

    The decorated function's first ``str`` parameter is used as the user
    message.  When ``streaming=True`` the wrapper becomes an async generator
    that yields tokens.

    Args:
        model: OpenAI model identifier.
        system_prompt: System message for the chat.
        temperature: Sampling temperature.
        max_tokens: Maximum tokens in the response.
        streaming: Whether to stream tokens.
        **extra_kwargs: Additional kwargs forwarded to the OpenAI client.

    Returns:
        A decorator that replaces the skill handler.

    Raises:
        ImportError: If the ``openai`` package is not installed.

    Example:
        @agent.skill("chat")
        @openai_skill(model="gpt-4o-mini")
        async def chat(message: str) -> str:
            ...
    """

    def decorator(func: Callable) -> Callable:
        if streaming:

            @functools.wraps(func)
            async def streaming_wrapper(**kwargs: Any):  # type: ignore[misc]
                try:
                    import openai
                except ImportError:
                    raise ImportError(
                        "OpenAI integration requires the 'openai' package. "
                        "Install it with: pip install a2a-lite[openai]"
                    )

                user_message = _extract_user_message(kwargs)
                client = openai.AsyncOpenAI(**extra_kwargs)
                messages = [
                    {"role": "system", "content": system_prompt},
                    {"role": "user", "content": user_message},
                ]
                create_kwargs: dict[str, Any] = {
                    "model": model,
                    "messages": messages,
                    "temperature": temperature,
                    "stream": True,
                }
                if max_tokens is not None:
                    create_kwargs["max_tokens"] = max_tokens

                stream = await client.chat.completions.create(**create_kwargs)
                async for chunk in stream:
                    delta = chunk.choices[0].delta if chunk.choices else None
                    if delta and delta.content:
                        yield delta.content

            return streaming_wrapper
        else:

            @functools.wraps(func)
            async def wrapper(**kwargs: Any) -> str:
                try:
                    import openai
                except ImportError:
                    raise ImportError(
                        "OpenAI integration requires the 'openai' package. "
                        "Install it with: pip install a2a-lite[openai]"
                    )

                user_message = _extract_user_message(kwargs)
                client = openai.AsyncOpenAI(**extra_kwargs)
                messages = [
                    {"role": "system", "content": system_prompt},
                    {"role": "user", "content": user_message},
                ]
                create_kwargs: dict[str, Any] = {
                    "model": model,
                    "messages": messages,
                    "temperature": temperature,
                }
                if max_tokens is not None:
                    create_kwargs["max_tokens"] = max_tokens

                response = await client.chat.completions.create(**create_kwargs)
                return response.choices[0].message.content or ""

            return wrapper

    return decorator


def anthropic_skill(
    model: str = "claude-sonnet-4-5-20250929",
    system_prompt: str = "You are a helpful assistant.",
    temperature: float = 0.7,
    max_tokens: int = 1024,
    streaming: bool = False,
    **extra_kwargs: Any,
) -> Callable:
    """Decorator that wraps a skill to call the Anthropic messages API.

    Args:
        model: Anthropic model identifier.
        system_prompt: System message for the conversation.
        temperature: Sampling temperature.
        max_tokens: Maximum tokens in the response.
        streaming: Whether to stream tokens.
        **extra_kwargs: Additional kwargs forwarded to the Anthropic client.

    Returns:
        A decorator that replaces the skill handler.

    Raises:
        ImportError: If the ``anthropic`` package is not installed.

    Example:
        @agent.skill("analyze")
        @anthropic_skill(model="claude-sonnet-4-5-20250929")
        async def analyze(text: str) -> str:
            ...
    """

    def decorator(func: Callable) -> Callable:
        if streaming:

            @functools.wraps(func)
            async def streaming_wrapper(**kwargs: Any):  # type: ignore[misc]
                try:
                    import anthropic
                except ImportError:
                    raise ImportError(
                        "Anthropic integration requires the 'anthropic' package. "
                        "Install it with: pip install a2a-lite[anthropic]"
                    )

                user_message = _extract_user_message(kwargs)
                client = anthropic.AsyncAnthropic(**extra_kwargs)

                async with client.messages.stream(
                    model=model,
                    system=system_prompt,
                    messages=[{"role": "user", "content": user_message}],
                    max_tokens=max_tokens,
                    temperature=temperature,
                ) as stream:
                    async for text in stream.text_stream:
                        yield text

            return streaming_wrapper
        else:

            @functools.wraps(func)
            async def wrapper(**kwargs: Any) -> str:
                try:
                    import anthropic
                except ImportError:
                    raise ImportError(
                        "Anthropic integration requires the 'anthropic' package. "
                        "Install it with: pip install a2a-lite[anthropic]"
                    )

                user_message = _extract_user_message(kwargs)
                client = anthropic.AsyncAnthropic(**extra_kwargs)
                response = await client.messages.create(
                    model=model,
                    system=system_prompt,
                    messages=[{"role": "user", "content": user_message}],
                    max_tokens=max_tokens,
                    temperature=temperature,
                )
                # Extract text from content blocks
                text_parts = [
                    block.text for block in response.content if hasattr(block, "text")
                ]
                return "".join(text_parts)

            return wrapper

    return decorator


def ollama_skill(
    model: str = "llama3.2",
    base_url: str = "http://localhost:11434",
    system_prompt: str = "You are a helpful assistant.",
    temperature: float = 0.7,
    streaming: bool = False,
) -> Callable:
    """Decorator that wraps a skill to call a local Ollama instance.

    Uses httpx (already a core dep) to call the Ollama HTTP API directly,
    so no additional packages are required.

    Args:
        model: Ollama model name.
        base_url: Ollama server URL.
        system_prompt: System message.
        temperature: Sampling temperature.
        streaming: Whether to stream tokens.

    Returns:
        A decorator that replaces the skill handler.

    Example:
        @agent.skill("local_chat")
        @ollama_skill(model="llama3.2")
        async def local_chat(message: str) -> str:
            ...
    """

    def decorator(func: Callable) -> Callable:
        if streaming:

            @functools.wraps(func)
            async def streaming_wrapper(**kwargs: Any):  # type: ignore[misc]
                import httpx

                user_message = _extract_user_message(kwargs)
                url = f"{base_url.rstrip('/')}/api/chat"
                payload = {
                    "model": model,
                    "messages": [
                        {"role": "system", "content": system_prompt},
                        {"role": "user", "content": user_message},
                    ],
                    "stream": True,
                    "options": {"temperature": temperature},
                }

                async with httpx.AsyncClient() as client:
                    async with client.stream(
                        "POST", url, json=payload, timeout=120.0
                    ) as response:
                        import json as _json

                        async for line in response.aiter_lines():
                            if line.strip():
                                data = _json.loads(line)
                                content = data.get("message", {}).get("content", "")
                                if content:
                                    yield content

            return streaming_wrapper
        else:

            @functools.wraps(func)
            async def wrapper(**kwargs: Any) -> str:
                import httpx

                user_message = _extract_user_message(kwargs)
                url = f"{base_url.rstrip('/')}/api/chat"
                payload = {
                    "model": model,
                    "messages": [
                        {"role": "system", "content": system_prompt},
                        {"role": "user", "content": user_message},
                    ],
                    "stream": False,
                    "options": {"temperature": temperature},
                }

                async with httpx.AsyncClient() as client:
                    response = await client.post(url, json=payload, timeout=120.0)
                    response.raise_for_status()
                    data = response.json()
                    return data.get("message", {}).get("content", "")

            return wrapper

    return decorator


def _extract_user_message(kwargs: dict[str, Any]) -> str:
    """Extract the user message from skill kwargs.

    Looks for common parameter names: message, text, query, prompt, input.
    Falls back to the first string value.

    Args:
        kwargs: The keyword arguments passed to the skill.

    Returns:
        The user message string.
    """
    for key in ("message", "text", "query", "prompt", "input"):
        if key in kwargs:
            return str(kwargs[key])
    # Fallback: use the first string value
    for value in kwargs.values():
        if isinstance(value, str):
            return value
    return str(next(iter(kwargs.values()), ""))

File: packages/typescript/src/llm.ts
/**
 * LLM integration for A2A Lite TypeScript.
 *
 * Provides higher-order functions that wrap skill handlers to call LLM APIs.
 * Uses dynamic import() so LLM SDKs are optional peer dependencies.
 *
 * Usage:
 *   import { openaiSkill, anthropicSkill } from 'a2a-lite';
 *
 *   agent.skill("chat", openaiSkill({ model: "gpt-4o-mini" }));
 *
 *   agent.skill("analyze", anthropicSkill({ model: "claude-sonnet-4-6" }));
 */

export interface OpenAISkillConfig {
  model?: string;
  systemPrompt?: string;
  temperature?: number;
  maxTokens?: number;
  streaming?: boolean;
}

export interface AnthropicSkillConfig {
  model?: string;
  systemPrompt?: string;
  temperature?: number;
  maxTokens?: number;
  streaming?: boolean;
}

type SkillHandler = (params: Record<string, unknown>) => Promise<unknown> | AsyncGenerator<unknown>;

/**
 * Returns a SkillHandler that calls the OpenAI chat completions API.
 *
 * Requires: npm install openai
 *
 * @example
 *   agent.skill("chat", openaiSkill({ model: "gpt-4o-mini", systemPrompt: "You are helpful." }));
 *   // or streaming:
 *   agent.skill("chat", { streaming: true }, openaiSkill({ model: "gpt-4o-mini", streaming: true }));
 */
export function openaiSkill(config: OpenAISkillConfig = {}): SkillHandler {
  const {
    model = 'gpt-4o-mini',
    systemPrompt = 'You are a helpful assistant.',
    temperature = 0.7,
    maxTokens,
    streaming = false,
  } = config;

  if (streaming) {
    return async function* (params: Record<string, unknown>) {
      let OpenAI: new () => {
        chat: { completions: { create(p: Record<string, unknown>): Promise<AsyncIterable<Record<string, unknown>>> } };
      };
      try {
        // @ts-ignore — openai is an optional peer dependency
        const mod = await import('openai');
        OpenAI = mod.default;
      } catch {
        throw new Error("OpenAI integration requires the 'openai' package. Install it with: npm install openai");
      }

      const client = new OpenAI();
      const userMessage = extractUserMessage(params);
      const messages = [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: userMessage },
      ];

      const createParams: Record<string, unknown> = { model, messages, temperature, stream: true };
      if (maxTokens != null) createParams.max_tokens = maxTokens;

      const stream = await client.chat.completions.create(createParams);
      for await (const chunk of stream) {
        const choices = (chunk as Record<string, unknown>).choices as Array<Record<string, unknown>> | undefined;
        const delta = choices?.[0]?.delta as Record<string, unknown> | undefined;
        const content = delta?.content as string | undefined;
        if (content) yield content;
      }
    };
  }

  return async (params: Record<string, unknown>): Promise<string> => {
    let OpenAI: new () => {
      chat: { completions: { create(p: Record<string, unknown>): Promise<Record<string, unknown>> } };
    };
    try {
      // @ts-ignore — openai is an optional peer dependency
      const mod = await import('openai');
      OpenAI = mod.default;
    } catch {
      throw new Error("OpenAI integration requires the 'openai' package. Install it with: npm install openai");
    }

    const client = new OpenAI();
    const userMessage = extractUserMessage(params);
    const messages = [
      { role: 'system', content: systemPrompt },
      { role: 'user', content: userMessage },
    ];

    const createParams: Record<string, unknown> = { model, messages, temperature };
    if (maxTokens != null) createParams.max_tokens = maxTokens;

    const response = await client.chat.completions.create(createParams);
    const choices = response.choices as Array<Record<string, unknown>> | undefined;
    const message = choices?.[0]?.message as Record<string, unknown> | undefined;
    return (message?.content as string) ?? '';
  };
}

/**
 * Returns a SkillHandler that calls the Anthropic messages API.
 *
 * Requires: npm install @anthropic-ai/sdk
 *
 * @example
 *   agent.skill("analyze", anthropicSkill({ model: "claude-sonnet-4-6" }));
 */
export function anthropicSkill(config: AnthropicSkillConfig = {}): SkillHandler {
  const {
    model = 'claude-sonnet-4-6',
    systemPrompt = 'You are a helpful assistant.',
    temperature = 0.7,
    maxTokens = 1024,
    streaming = false,
  } = config;

  if (streaming) {
    return async function* (params: Record<string, unknown>) {
      let Anthropic: new () => {
        messages: { stream(p: Record<string, unknown>): AsyncIterable<Record<string, unknown>> };
      };
      try {
        // @ts-ignore — @anthropic-ai/sdk is an optional peer dependency
        const mod = await import('@anthropic-ai/sdk');
        Anthropic = mod.default;
      } catch {
        throw new Error(
          "Anthropic integration requires the '@anthropic-ai/sdk' package. Install it with: npm install @anthropic-ai/sdk",
        );
      }

      const client = new Anthropic();
      const userMessage = extractUserMessage(params);

      const stream = client.messages.stream({
        model,
        system: systemPrompt,
        messages: [{ role: 'user', content: userMessage }],
        max_tokens: maxTokens,
        temperature,
      });

      for await (const event of stream) {
        const delta = (event as Record<string, unknown>).delta as Record<string, unknown> | undefined;
        const text = delta?.text as string | undefined;
        if (text) yield text;
      }
    };
  }

  return async (params: Record<string, unknown>): Promise<string> => {
    let Anthropic: new () => { messages: { create(p: Record<string, unknown>): Promise<Record<string, unknown>> } };
    try {
      // @ts-ignore — @anthropic-ai/sdk is an optional peer dependency
      const mod = await import('@anthropic-ai/sdk');
      Anthropic = mod.default;
    } catch {
      throw new Error(
        "Anthropic integration requires the '@anthropic-ai/sdk' package. Install it with: npm install @anthropic-ai/sdk",
      );
    }

    const client = new Anthropic();
    const userMessage = extractUserMessage(params);

    const response = await client.messages.create({
      model,
      system: systemPrompt,
      messages: [{ role: 'user', content: userMessage }],
      max_tokens: maxTokens,
      temperature,
    });

    const content = response.content as Array<Record<string, unknown>>;
    return content
      .filter((block) => block.type === 'text')
      .map((block) => block.text as string)
      .join('');
  };
}

export interface OllamaSkillConfig {
  model?: string;
  baseUrl?: string;
  systemPrompt?: string;
  temperature?: number;
  streaming?: boolean;
}

/**
 * Returns a SkillHandler that calls a local Ollama instance.
 *
 * No extra packages needed — uses native fetch (Node 18+).
 *
 * @example
 *   agent.skill("chat", ollamaSkill({ model: "llama3.2" }));
 */
export function ollamaSkill(config: OllamaSkillConfig = {}): SkillHandler {
  const {
    model = 'llama3.2',
    baseUrl = 'http://localhost:11434',
    systemPrompt = 'You are a helpful assistant.',
    temperature = 0.7,
    streaming = false,
  } = config;

  const url = `${baseUrl.replace(/\/$/, '')}/api/chat`;
  const buildPayload = (userMessage: string, stream: boolean) => ({
    model,
    messages: [
      { role: 'system', content: systemPrompt },
      { role: 'user', content: userMessage },
    ],
    stream,
    options: { temperature },
  });

  if (streaming) {
    return async function* (params: Record<string, unknown>) {
      const userMessage = extractUserMessage(params);
      const response = await fetch(url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(buildPayload(userMessage, true)),
      });
      if (!response.ok || !response.body) {
        throw new Error(`Ollama returned HTTP ${response.status}`);
      }
      const reader = response.body.getReader();
      const decoder = new TextDecoder();
      let buffer = '';
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;
        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split('\n');
        buffer = lines.pop() ?? '';
        for (const line of lines) {
          if (!line.trim()) continue;
          try {
            const data = JSON.parse(line) as Record<string, unknown>;
            const content = (data.message as Record<string, unknown>)?.content as string;
            if (content) yield content;
          } catch {
            /* skip malformed lines */
          }
        }
      }
    };
  }

  return async (params: Record<string, unknown>): Promise<string> => {
    const userMessage = extractUserMessage(params);
    const response = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(buildPayload(userMessage, false)),
    });
    if (!response.ok) throw new Error(`Ollama returned HTTP ${response.status}`);
    const data = (await response.json()) as Record<string, unknown>;
    return ((data.message as Record<string, unknown>)?.content as string) ?? '';
  };
}

export interface BedrockSkillConfig {
  model?: string;
  region?: string;
  systemPrompt?: string;
  temperature?: number;
  maxTokens?: number;
  streaming?: boolean;
}

/**
 * Returns a SkillHandler that calls the AWS Bedrock Converse API.
 *
 * Works with any Bedrock model (Claude, Llama, Mistral, Nova, etc.)
 *
 * Requires: npm install @aws-sdk/client-bedrock-runtime
 *
 * @example
 *   agent.skill("chat", bedrockSkill({ model: "anthropic.claude-3-haiku-20240307-v1:0" }));
 */
export function bedrockSkill(config: BedrockSkillConfig = {}): SkillHandler {
  const {
    model = 'anthropic.claude-3-haiku-20240307-v1:0',
    region = 'us-east-1',
    systemPrompt = 'You are a helpful assistant.',
    temperature = 0.7,
    maxTokens = 1024,
    streaming = false,
  } = config;

  // Optional peer dependency: @aws-sdk/client-bedrock-runtime is not installed at compile time.
  // We use a structural type that matches the subset of BedrockRuntimeClient we need.
  interface BedrockClient {
    send(command: unknown): Promise<Record<string, unknown>>;
  }

  async function getClient(): Promise<BedrockClient> {
    try {
      // @ts-ignore — @aws-sdk/client-bedrock-runtime is an optional peer dependency
      const { BedrockRuntimeClient } = await import('@aws-sdk/client-bedrock-runtime');
      return new BedrockRuntimeClient({ region }) as BedrockClient;
    } catch {
      throw new Error(
        "Bedrock integration requires '@aws-sdk/client-bedrock-runtime'. " +
          'Install it with: npm install @aws-sdk/client-bedrock-runtime',
      );
    }
  }

  if (streaming) {
    return async function* (params: Record<string, unknown>) {
      // @ts-ignore — @aws-sdk/client-bedrock-runtime is an optional peer dependency
      const { ConverseStreamCommand } = await import('@aws-sdk/client-bedrock-runtime').catch(() => {
        throw new Error(
          "Bedrock integration requires '@aws-sdk/client-bedrock-runtime'. " +
            'Install it with: npm install @aws-sdk/client-bedrock-runtime',
        );
      });
      const client = await getClient();
      const userMessage = extractUserMessage(params);
      const command = new ConverseStreamCommand({
        modelId: model,
        system: [{ text: systemPrompt }],
        messages: [{ role: 'user', content: [{ text: userMessage }] }],
        inferenceConfig: { maxTokens, temperature },
      });
      const response = await client.send(command);
      const stream = response['stream'] as AsyncIterable<Record<string, unknown>> | undefined;
      for await (const event of stream ?? []) {
        const delta = (event['contentBlockDelta'] as Record<string, unknown> | undefined)?.['delta'] as
          | Record<string, unknown>
          | undefined;
        const text = delta?.['text'] as string | undefined;
        if (text) yield text;
      }
    };
  }

  return async (params: Record<string, unknown>): Promise<string> => {
    // @ts-ignore — @aws-sdk/client-bedrock-runtime is an optional peer dependency
    const { ConverseCommand } = await import('@aws-sdk/client-bedrock-runtime').catch(() => {
      throw new Error(
        "Bedrock integration requires '@aws-sdk/client-bedrock-runtime'. " +
          'Install it with: npm install @aws-sdk/client-bedrock-runtime',
      );
    });
    const client = await getClient();
    const userMessage = extractUserMessage(params);
    const command = new ConverseCommand({
      modelId: model,
      system: [{ text: systemPrompt }],
      messages: [{ role: 'user', content: [{ text: userMessage }] }],
      inferenceConfig: { maxTokens, temperature },
    });
    const response = await client.send(command);
    const output = response['output'] as Record<string, unknown> | undefined;
    const message = output?.['message'] as Record<string, unknown> | undefined;
    const content = (message?.['content'] as Array<Record<string, unknown>> | undefined) ?? [];
    return content
      .filter((b: Record<string, unknown>) => b['text'] != null)
      .map((b: Record<string, unknown>) => b['text'] as string)
      .join('');
  };
}

function extractUserMessage(params: Record<string, unknown>): string {
  for (const key of ['message', 'text', 'query', 'prompt', 'input']) {
    if (key in params) return String(params[key]);
  }
  const first = Object.values(params)[0];
  return first != null ? String(first) : '';
}
